src/lib.rs

```diff
@@ -78,70 +82,65 @@ impl Default for AttestationAgent {
-        let kbc_instance = (instantiate_func)(kbs_uri);
-        self.register_instance(kbc_name, kbc_instance);
+    fn instantiate_kbc(&mut self, kbc_name: &str, kbs_uri: &str) -> Result<()> {
+        let instantiate_func = self.kbc_module_list.get_func(kbc_name)?;
+        let kbc_instance = (instantiate_func)(kbs_uri.to_string());
+        self.register_instance(kbc_name.to_string(), kbc_instance);
         Ok(())
     }
 
     #[allow(dead_code)]
     fn check(&self, kbc_name: String) -> Result<KbcCheckInfo> {
-        let kbc_instance = self
-            .kbc_instance_map
+        self.kbc_instance_map
             .get(&kbc_name)
-            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?;
-        let check_info: KbcCheckInfo = kbc_instance.check()?;
-        Ok(check_info)
+            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?
+            .check()
     }
 }
 
 #[async_trait]
 impl AttestationAPIs for AttestationAgent {
     async fn decrypt_image_layer_annotation(
         &mut self,
-        kbc_name: String,
-        kbs_uri: String,
-        annotation: String,
+        kbc_name: &str,
+        kbs_uri: &str,
+        annotation: &str,
     ) -> Result<Vec<u8>> {
-        if self.kbc_instance_map.get_mut(&kbc_name).is_none() {
-            self.instantiate_kbc(kbc_name.clone(), kbs_uri)?;
+        if !self.kbc_instance_map.contains_key(kbc_name) {
+            self.instantiate_kbc(kbc_name, kbs_uri)?;
         }
-        let kbc_instance = self
-            .kbc_instance_map
-            .get_mut(&kbc_name)
-            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?;
-        let plain_payload = kbc_instance.decrypt_payload(&annotation).await?;
-        Ok(plain_payload)
+        self.kbc_instance_map
+            .get_mut(kbc_name)
+            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?
+            .decrypt_payload(&annotation)
+            .await
     }
 
     async fn download_confidential_resource(
         &mut self,
-        kbc_name: String,
-        kbs_uri: String,
-        resource_description: String,
+        kbc_name: &str,
+        kbs_uri: &str,
+        resource_description: &str,
     ) -> Result<Vec<u8>> {
-        if self.kbc_instance_map.get_mut(&kbc_name).is_none() {
-            self.instantiate_kbc(kbc_name.clone(), kbs_uri)?;
+        if !self.kbc_instance_map.contains_key(kbc_name) {
+            self.instantiate_kbc(kbc_name, kbs_uri)?;
         }
-        let kbc_instance = self
-            .kbc_instance_map
-            .get_mut(&kbc_name)
-            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?;
-        let resource = kbc_instance.get_resource(resource_description).await?;
-        Ok(resource)
+        self.kbc_instance_map
+            .get_mut(kbc_name)
+            .ok_or_else(|| anyhow!("The KBC instance does not existing!"))?
+            .get_resource(resource_description)
+            .await
```

Comment:
Good suggestion, fixed:)

Reply:
Just have a try and found `.get_mut(kbc_name)` will return `None` instead of `Err(e)`.
We can try HashMap::entry().ok_or_insert(), but that change is a little big:)

Reply:
ah, you're right. OK, just leave as-is (the `ok_or_else` line "converts" the None to an Err).